core/CronArchive/QueueConsumer.php

```diff
@@ -311,6 +315,66 @@ public function getNextArchivesToProcess()
         return $archivesToProcess;
     }
 
+    // public for tests
+    public function repairInvalidationsIfNeeded($archiveToProcess)
+    {
+        $table = Common::prefixTable('archive_invalidations');
+
+        $bind = [
+            $archiveToProcess['idsite'],
+            $archiveToProcess['name'],
+            $archiveToProcess['period'],
+            $archiveToProcess['date1'],
+            $archiveToProcess['date2'],
+        ];
+
+        $reportClause = '';
+        if (!empty($archiveToProcess['report'])) {
+            $reportClause = " AND report = ?";
+            $bind[] = $archiveToProcess['report'];
+        }
+
+        $sql = "SELECT DISTINCT period FROM `$table` WHERE idsite = ? AND name = ? AND period > ? AND ? >= date1 AND date2 >= ? $reportClause";
+
+        $higherPeriods = Db::fetchAll($sql, $bind);
+        $higherPeriods = array_column($higherPeriods, 'period');
+        $higherPeriods = array_flip($higherPeriods);
+
+        $invalidationsToInsert = [];
+        foreach (Piwik::$idPeriods as $label => $id) {
+            // lower period than the one we're processing or range, don't care
+            if ($id <= $archiveToProcess['period'] || $label == 'range') {
+                continue;
+            }
+
+            if (isset($higherPeriods[$id])) { // period exists in table
+                continue;
+            }
+
+            // archive is for week that is over two months, we don't need to care about the month
+            if ($label == 'month'
```

Comment:
not sure why we don't need to care about this case? There might be still issues eg when someone imports data in the past etc?

Reply:
It's because the week will be on two months, so it won't actually be used in archiving either of the parent months.

There could be the case where there was a day that was archived, then for some reason the month/year disappear, and it's just the week. Then I guess we would need new month/year re-archives, but wouldn't be able to.

It seemed a waste to have to re-archive both months, but I can add it if necessary.

Reply:
Got it now 👍  We only need to archive the correct month